ang/civicase/ActivityCard.js

```diff
@@ -1,56 +1,103 @@
-(function(angular, $, _) {
+(function (angular, $, _) {
+  var module = angular.module('civicase');
 
-  function activityCard($scope, getActivityFeedUrl, dialogService, templateExists) {
+  module.directive('caseActivityCard', function () {
+    return {
+      restrict: 'A',
+      templateUrl: '~/civicase/ActivityCard.html',
+      controller: caseActivityCardController,
+      scope: {
+        activity: '=caseActivityCard',
+        refresh: '=refreshCallback',
+        editActivityUrl: '=?editActivityUrl'
+      }
+    };
+  });
+
+  function caseActivityCardController ($scope, getActivityFeedUrl, dialogService, templateExists, crmApi) {
     var ts = $scope.ts = CRM.ts('civicase');
-    $scope.CRM = CRM;
     $scope.activityFeedUrl = getActivityFeedUrl;
+    $scope.CRM = CRM;
     $scope.templateExists = templateExists;
 
-    $scope.isActivityEditable = function(activity) {
+    // console.log($scope.activity.category);
+
+    $scope.isActivityEditable = function (activity) {
       var type = CRM.civicase.activityTypes[activity.activity_type_id].name;
+
       return (type !== 'Email' && type !== 'Print PDF Letter') && $scope.editActivityUrl;
     };
 
-    $scope.markCompleted = function(act) {
-      $scope.refresh([['Activity', 'create', {id: act.id, status_id: act.is_completed ? 'Scheduled' : 'Completed'}]]);
+    /**
+     * Mark an activity as complete
+     *
+     * @param {object} activity
+     */
+    $scope.markCompleted = function (activity) {
+      return crmApi([['Activity', 'create', {id: activity.id, status_id: activity.is_completed ? 'Scheduled' : 'Completed'}]])
```

Comment:
Should we create a service out of it. It will make it modular. Also, I feel any API calls should be done through a service. Your thoughts on this?

Reply:
`crmApi` is already an angular factory defined in civicrm core. 
https://github.com/civicrm/civicrm-core/blob/master/ang/crmUtil.js#L13

So I am using the same. Initially I too thought about creating a separate service, but did not do
1. We should not change the architecture too much. And as core is already providing a factory for this, we should not change it now.